src/plugins/compressors/sz_threadsafe_plugin.cc

```diff
@@ -0,0 +1,424 @@
+#include <algorithm>
+#include <iterator>
+#include <memory>
+#include <sstream>
+#include <cstdlib>
+
+#include <sz/sz.h>
+#if HAVE_WRITESTATS
+#include <sz/sz_threadsafe_stats.h>
+#endif
+
+#include "libpressio_ext/cpp/data.h"
+#include "libpressio_ext/cpp/compressor.h"
+#include "libpressio_ext/cpp/pressio.h"
+#include "libpressio_ext/cpp/options.h"
+#include "std_compat/std_compat.h"
+#include "pressio_data.h"
+#include "pressio_compressor.h"
+#include "pressio_options.h"
+#include "pressio_option.h"
+#include "std_compat/memory.h"
+
+#define PRESSIO_SZ_VERSION_GREATEREQ(major, minor, build, revision) \
+   (SZ_VER_MAJOR > major || \
+   (SZ_VER_MAJOR == major && SZ_VER_MINOR > minor) ||                                  \
+   (SZ_VER_MAJOR == major && SZ_VER_MINOR == minor && SZ_VER_BUILD > build) || \
+   (SZ_VER_MAJOR == major && SZ_VER_MINOR == minor && SZ_VER_BUILD == build && SZ_VER_REVISION >= revision))
+
+namespace {
+  struct iless {
+    bool operator()(std::string lhs, std::string rhs) const {
+      std::transform(std::begin(lhs), std::end(lhs), std::begin(lhs), [](unsigned char c){return std::tolower(c);});
+      std::transform(std::begin(rhs), std::end(rhs), std::begin(rhs), [](unsigned char c){return std::tolower(c);});
+      return lhs < rhs;
+    }
+  };
+  static std::map<std::string, int, iless> const sz_mode_str_to_code {
+    {"abs", ABS},
+    {"rel", REL},
+    {"vr_rel", REL},
+    {"abs_and_rel", ABS_AND_REL},
+    {"abs_or_rel", ABS_OR_REL},
+    {"psnr", PSNR},
+#if PRESSIO_SZ_VERSION_GREATEREQ(2,1,8,3)
+    {"norm", NORM},
+#endif
+    {"pw_rel", PW_REL},
+    {"abs_or_pw_rel", ABS_OR_PW_REL},
+    {"abs_and_pw_rel", ABS_AND_PW_REL},
+    {"rel_or_pw_rel", REL_OR_PW_REL},
+    {"rel_and_pw_rel", REL_AND_PW_REL},
+  };
+}
```

Comment:
This code should be refactored into a common component.  It didn't trigger an ODR violation before because of the anonymous namespace and being a macro, but it's cleaner to put this in a common place.